### PR TITLE
feat: catalog source pod scheduling config overrides

### PR DESCRIPTION
**Pull request by perdasilva**
**Description of the change:**
The CatalogSource CRD was recently [updated](https://github.com/operator-framework/api/pull/173) to carry an additional and optional attribute `grpcPodConfig` as an envelope for pod spec overrides that we may want to give to the users, starting with: `tolerations`, `nodeSelector`, and `priorityClassName`. For instance:

```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
...
# optional
grpcPodConfig:
  # optional
  priorityClassName: https://github.com/operator-framework/api/pull/173
  #optional
  tolerations:
    - key: some
      operator: Equal
      value: value
      effect: NoSchedule
  # optional
  nodeSelector: 
     kubernetes.io/os: linux
     some: tag
```

**Motivation for the change:**
We want to give cluster admins a way to force catalog source pods to be scheduled on specific infrastructure (i.e. nodes)

**Questions**

**Does it make sense to expose `priorityClassName` and not `priority`?**

`priorityClassName` seems to be used with special (custom) priority classes and `system-node-critical` and `system-cluster-critical` (which I think are reserved and very high priority). I think it should be fine to **not** expose the `priority` knob as well. I think `priorityClassName` covers most cases where an admin would want to affect priority. But, I'd be keen to have this assumption checked =)

**Should we update the api and use `*string` for `PriorityClassName`**

Since we want to override settings, it could make sense to differentiate between not set and empty. A priorityClassName of `""` says to use the default priority. We don't set our own priorityClassName in our catalog source pod definition. So, the only issue here is that in case we decide to do that (which I don't think is likely), the user will not be able to unset it, i.e. override it to be `""`. But, even though unlikely, if we have the chance and it is quick, why not "do it right"? (if you guys also agree that seems the most correctest)

**NOTE:** This PR will eventually also introduce `vendor` changes. For reviewer sanity, I'll try to keep those off the PR for as long as possible

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by perdasilva** [comment]
holding until api changes are finalized and merged

**Review by tylerslaton — COMMENTED** [review]
Looks good to me.

non-block question: Do we need to write an E2E test that validates that the changes to the pod spec are actually being utilized as expected? Or is that something that is owned by some other k8s API?

**Comment by perdasilva** [comment]
> non-block question: Do we need to write an E2E test that validates that the changes to the pod spec are actually being utilized as expected? Or is that something that is owned by some other k8s API?

Yeah, I'll def add some e2e tests!

File: pkg/controller/registry/reconciler/reconciler.go
Comment (njhale) [inline pkg/controller/registry/reconciler/reconciler.go:137]: nit: the last right-bracket on this line matches `ExecAction{` -- is this change a typo?

File: pkg/controller/registry/reconciler/reconciler.go
Comment (njhale) [inline pkg/controller/registry/reconciler/reconciler.go:184]: @perdasilva Now that I'm reading it, I think you're right to suggest we make this a pointer -- it leaves the option open for OLM to have some default that's not `""` and for users to opt into the original.

**Review by njhale — COMMENTED** [review]
> NOTE: This PR will eventually also introduce vendor changes. For reviewer sanity, I'll try to keep those off the PR for as long as possible

I think it's sufficiently clean to include vendor changes in a separate commit, that way we can still view the meaningful diffs separately and don't need to re-lgtm later.

Other than that, this is looking good! I think we just need an e2e test to make sure OLM respects these settings and updates existing catalog pods when changes are detected.

File: deploy/chart/crds/0000_50_olm_00-catalogsources.crd.yaml
Comment (perdasilva) [reply deploy/chart/crds/0000_50_olm_00-catalogsources.crd.yaml:248]: it may have been me - I think this will be fixed in the next iteration when I include all the vendored changes properly

File: pkg/controller/registry/reconciler/reconciler.go
Comment (perdasilva) [reply pkg/controller/registry/reconciler/reconciler.go:137]: might have come from goland format or make format. I'll :eyes:

File: pkg/controller/registry/reconciler/reconciler.go
Comment (timflannagan) [reply pkg/controller/registry/reconciler/reconciler.go:184]: Any thoughts on how we should derive configuration when the `operatorframework.io/priorityclass` annotation has been specified on a CatalogSource: https://github.com/operator-framework/operator-lifecycle-manager/blob/4fc476fc520d4d595368ffc8e4c032c4d34dd796/pkg/controller/registry/reconciler/reconciler.go#L27? I'm not entirely sure if that hacky way of propagating configuration to the underlying Pod violates any API contract? It might also be entirely reasonable to point users towards these new configuration fields in the CatalogSource entirely, but it's unknown how many users were relying on that behavior.

Comment (gallettilance) [reply pkg/controller/registry/reconciler/reconciler.go:184]: this change seems to be backward compatible. If I had a priorityclass set on the catalogsource and now I add a config that contains a priority class, it's still [the former that is propagated through](https://github.com/operator-framework/operator-lifecycle-manager/pull/2512/files#diff-fffdeef1fc140a5dc5dc92dda323f567a6e46fc2ecbb0b91ba907acd02bde50dL173-L175).

**Review by gallettilance — APPROVED** [review]
lgtm

File: pkg/controller/registry/reconciler/reconciler.go
Comment (perdasilva) [reply pkg/controller/registry/reconciler/reconciler.go]: @timflannagan I've added a test to ensure that the annotation takes precedence. We'll handle this in documentation as well. We should probably deprecate the annotation tho.

File: pkg/controller/registry/reconciler/reconciler_test.go
Comment (timflannagan) [inline pkg/controller/registry/reconciler/reconciler_test.go:243]: :+1:

File: test/e2e/catsrc_pod_config_e2e_test.go
Comment (timflannagan) [inline test/e2e/catsrc_pod_config_e2e_test.go:16]: nit: reminder to remove this

**Comment by perdasilva** [comment]
> non-block question: Do we need to write an E2E test that validates that the changes to the pod spec are actually being utilized as expected? Or is that something that is owned by some other k8s API?

@tylerslaton I thought about it, but yeah. That's the conclusion I reached. As long as its setting the right things in the right places the rest is up to kube

File: test/e2e/ctx/provisioner_kind.go
Comment (perdasilva) [reply test/e2e/ctx/provisioner_kind.go:86]: I'll update to use the context logger! Nice one! I think this is useful output (even though you can just export the kubeconfig - also didn't know about _that_ - so thanks for that too!

File: pkg/controller/registry/reconciler/reconciler.go
Comment (perdasilva) [reply pkg/controller/registry/reconciler/reconciler.go:137]: I think that's correct - it may have been just a format change.

File: pkg/controller/registry/reconciler/reconciler.go
Comment (anik120) [inline pkg/controller/registry/reconciler/reconciler.go:181]: non blocking nit: why is copying the map necessary here? Why not make `grpcPodConfig.NodeSelector` and `pod.Spec.NodeSelector` point to the same thing?

File: test/e2e/catsrc_pod_config_e2e_test.go
Comment (anik120) [inline test/e2e/catsrc_pod_config_e2e_test.go:75]: Another non-blocking nit: I'll be surprised if this isn't already there in a `utils`, but this looks like code that'll be repeated fairly often. Something along the line of `deleteObj(client, name, namespace)` func could help reduce the redundancy.

**Review by anik120 — CHANGES_REQUESTED** [review]
Looks good mostly @perdasilva , had a couple of non-blocking questions otherwise this looks good to go.

File: pkg/controller/registry/reconciler/reconciler.go
Comment (perdasilva) [reply pkg/controller/registry/reconciler/reconciler.go:181]: I thought I'd be defensive here because the nodeselector map is passed by reference. If possible (thought probably unlikely) that you could have out of band changes make it into the pod spec (e.g. if something else changes the map). So, I thought it safer to copy it over and avoid possible mysterious bugs.

File: test/e2e/catsrc_pod_config_e2e_test.go
Comment (perdasilva) [reply test/e2e/catsrc_pod_config_e2e_test.go:75]: This is a great point - I will check if I can find anything

File: pkg/controller/registry/reconciler/reconciler.go
Comment (anik120) [reply pkg/controller/registry/reconciler/reconciler.go:181]: okay that's fair

**Pull request merged**
